### PR TITLE
Don't allow recomputing a node that *must* be materialized in the backwards pass

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90896

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
This is a reland right? Giving you the stamp, though it would be nice to know what the delta from last time is.

File: torch/_functorch/partitioners.py [hunk 1]
```diff
@@ -360,14 +361,17 @@ def is_tensor_node(x):
...
     if compiler == "inductor":
         default_recomputable_ops += [prims.div, prims.convert_element_type, aten.clone, aten._to_copy, aten.full_like, prims.var, prims.sum, aten.var, aten.std, prims.broadcast_in_dim, aten.select, aten.permute, aten._unsafe_view, aten.view, aten.expand, aten.slice, aten.reshape, aten.broadcast_tensors, aten.scalar_tensor, aten.ones, aten.new_zeros, aten.lift_fresh_copy, aten.arange, aten.triu, aten.var_mean, aten.isinf, aten.any, aten.full, aten.as_strided, aten.zeros, aten.argmax, aten.maximum]  # noqa: E501
+        view_ops += [aten.view, aten.slice, aten.permute, aten.t, prims.broadcast_in_dim, aten.expand, aten.as_strided]
         # Natalia said that we should allow recomputing indexing :)
         default_recomputable_ops += [aten.index]
+    default_recomputable_ops += view_ops
 
-        # add more generally ?
-        default_recomputable_ops += pointwise_ops()
+    # add more generally ?
```
Comment (albanD) [inline torch/_functorch/partitioners.py:373]: I guess this comment can go? :)

File: torch/_functorch/config.py [hunk 2]
```diff
@@ -32,6 +32,12 @@
 
 static_weight_shapes = True
 
+# Applies CSE to the graph before partitioning
+cse = True
+
+# Restricts the amount of computation AOTAutograd can do.
+max_dist_from_bw = 5
```
Comment (albanD) [inline torch/_functorch/config.py:39]: This is bumped from 3 to 5. What is the impact on all the benchmarks?

File: torch/_functorch/partitioners.py [hunk 3]
```diff
@@ -389,6 +393,18 @@ def is_tensor_node(x):
 
     AGGRESSIVE_RECOMPUTATION = False
 
+    def is_materialized_backwards(node):
```
Comment (albanD) [inline torch/_functorch/partitioners.py:396]: nit: it took me a while to understand that this means: only computed during backward and not during forward. A small comment would be nice.

File: test/inductor/test_torchinductor.py [hunk 4]
```diff
@@ -5663,7 +5663,7 @@ def forward(self, x):
                         opt_fn = torch._dynamo.optimize("inductor")(m)
                         same(m(x), opt_fn(x))
                         if simdlen != 1:
-                            assert metrics.generated_cpp_vec_kernel_count == 7
+                            assert metrics.generated_cpp_vec_kernel_count == 6
```
Comment (Chillee) [inline test/inductor/test_torchinductor.py:5666]: @jgong5 This PR causes us to generate *less* kernels than we used to, which seems like a good thing :)

**Comment by eellison** [comment]
With https://github.com/pytorch/pytorch/pull/92863 applied, this commit regresses memory compression for 

`python benchmarks/dynamo/timm_models.py  --float32 --training --only=mobilevit_s --performance --inductor --disable-cudagraphs` from 1.40x -> 1.00x (and from 1.19x -> 1.00 without it applied). Do we have any performance movement that warrants that change ? Should we revert ?

cc @ngimel

**Comment by Chillee** [comment]
@eellison does it improve performance on that model?

**Comment by eellison** [comment]
gives a 6% speedup at a 40% compression ratio cost. 

don't we still want to recompute a node if it is produced by a node that doesn't need to be materialized ? this is stopping recomputation if it cant be fused with consumer, but it might be able to be fused with producer.

**Comment by ngimel** [comment]
Even if it's fused with producer, it's still an extra output, but I don't know how to balance memory with perf in this case, 40% indeed seems like a lot.

**Comment by eellison** [comment]
If it’s fused with a producer and it’s the producers only use, shouldn’t be an extra output I think.

**Comment by Chillee** [comment]
@eellison So, 2 comments.

1. I think, by default, we should always prioritize performance over memory, particularly when the memory hit isn't a regression over eager. That being said, there are many cases (particularly in distributed) where memory is a premium, so i think it's important that users have more fine-grained control of how much memory they're willing to sacrifice for performance (whether it's through composing with existing activation checkpointing APIs, or a more compiler-centric approach).
2. https://github.com/pytorch/pytorch/pull/89171#discussion_r1026804856 explains it a bit more. We can chat more offline if the explanation is still confusing.

**Pull request closed without merging**